### ISSUE TITLE
ID fields are too short for CUID in Prisma/Postgres workflow

### ISSUE THREAD

**Issue by louisrli** [prisma/upgrade#34]
I haven't investigated this fully, but I want to note that I've done nothing manually except:
* Had a pretty standard Prisma 1 postgres setup by following the tutorial
* Ran prisma upgrade and introspect

In the end, my Prisma 2 schema has fields like these: 
```
id               String            @default(cuid()) @id
```

I did not write the above manually, I'm not sure if it came from introspect or upgrade, but in the end I suppose it's related to the upgrade flow.

Anyway, in Postgres, this gives us a table like this. Again, nothing manual from me here, I believe that Prisma 1 created these `VARCHAR(25)`s if I had to take a guess (unless prisma-upgrade modified the ID fields in some way).

```
prisma=# \d default$default."BundleSection";
                          Table "default$default.BundleSection"
      Column      |                 Type                 | Collation | Nullable | Default
------------------+--------------------------------------+-----------+----------+---------
 id               | character varying(25)                |           | not null |
```

Unfortunately, `cuid()` can be longer than 25 characters, as seen here: https://github.com/ericelliott/cuid/issues/164#issuecomment-516637547

I'm not familiar with the exact order of operations in which the prisma1, prisma upgrade and introspect workflow so unfortunately I can't be of further help in diagnosing exactly where this was problem was, but I suppose what comes to mind is that:
* If Prisma 1 was the one that set `VARCHAR(25)`, prisma-upgrade should be cognizant of this if it chooses to default to `cuid` (since this breaks backwards compatibility with whatever id scheme prisma had)
* If Prisma upgrade was the one that set `VARCHAR(25)`, it should perhaps set a safer number

**Comment by louisrli** [comment]
Forgot to link to related issue in prisma:
https://github.com/prisma/prisma/issues/2730#issuecomment-652865551

**Comment by matthewmueller** [comment]
Hi @louisrli thanks for pointing this out! cuid's being longer than 25 characters came as a surprise to a couple of us internally 😅

How did this problem manifest itself for you? 

- Was it when you tried to run the prisma upgrade tool, one of the SQL commands failed? If so, which command was it?
- Was it after you completed an upgrade, but then you went to run your application, your Prisma 2 client failed? If so, what was the error message?

Or perhaps it's a different problem altogether?

**Comment by janpio** [comment]
https://github.com/prisma/prisma/issues/2730#issue-637815088 has information that it was in `update()`

**Comment by matthewmueller** [comment]
Okay so it sounds like you made it through the upgrade tool successfully, but now you went to run your application and it failed with:

```
Invalid `prisma.studyGuide.update()` invocation:


  The provided value for the column is too long for the column's type. Column: <unknown>
    at PrismaClientFetcher.request (/Users/louis/branches/srs/server/node_modules/@prisma/client/src/runtime/getPrismaClient.ts:762:15)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at main (/Users/louis/branches/srs/server/prisma.bak/spreadsheetseed.ts:104:5) {
  code: 'P2000',
  meta: { column_name: '<unknown>' }
}
```

- Do you mind sharing the full `update()` command? 
- Are you updating an existing record? If so, please share the ID of that record.

**Comment by louisrli** [comment]
* I'm not updating an existing record
* The full update command is constructed programmatically through a series of long and nested creates. Basically, I'm running into this issue in a seed script for seeding my local/staging databases; the same script worked in Prisma 1.

I figured out that it was an ID issue and not any other scalar column because, although the same field was deterministically failing on insertion at first, I basically did something like:

```
if (some_field_value === 'an identifier for the failing field') {
   continue
} 
```

And I found that some other random insertion would start failing.

I also found that if I isolated the insertions in a single run of the script to 1/8th of the insertions, then there would be no failure in the column being too long.

tldr basically I'm doing a large number of inserts to seed a database with some base data, and when the number of insertions in a single run of the script is large, it runs into this error

**Comment by matthewmueller** [comment]
Thanks for the info, so it appears that it's related to the Prisma Engine generating more than 25 character IDs in some cases during a create. I've passed this on to the team to have a closer look.

**Comment by janpio** [comment]
Hey @louisrli, if you are in a position to just rerun this, could you please configure your Prisma Client to log the actual SQL queries executed please? https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/logging#overview It would be helpful to have an example of such an ID that is causing this to make sure we are not hunting a ghost.

**Comment by louisrli** [comment]
Is there a way to view the parameter values?

`"$default$default"."RawContent"."content"` is a `TEXT` field here, by the way, so I suppose the only candidate for failure is the id.

```
prisma:query BEGIN
prisma:query SELECT "default$default"."BundleTopic"."id" FROM "default$default"."BundleTopic" WHERE "default$default"."BundleTopic"."id" = $1 OFFSET $2
prisma:query INSERT INTO "default$default"."BundleSection" ("id","title","topic") VALUES ($1,$2,$3) RETURNING "default$default"."BundleSection"."id"
prisma:query INSERT INTO "default$default"."RawContent" ("content","id") VALUES ($1,$2) RETURNING "default$default"."RawContent"."id"
prisma:query ROLLBACK
PrismaClientKnownRequestError:
Invalid `prisma.bundleSection.create()` invocation in
/Users/louis/branches/srs/server/prisma/spreadsheet/spec-to-prisma.ts:161:36

  157   connect: {
  158     id: newTopic!.id,
  159   },
  160 };
→ 161 await prisma.bundleSection.create(
  The provided value for the column is too long for the column's type. Column: <unknown>
    at PrismaClientFetcher.request (/Users/louis/branches/srs/server/node_modules/@prisma/client/src/runtime/getPrismaClient.ts:762:15)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at Object.insertSubjectWithPrisma (/Users/louis/branches/srs/server/prisma/spreadsheet/spec-to-prisma.ts:161:9)
    at main (/Users/louis/branches/srs/server/prisma/spreadsheetseed.ts:78:5) {
  code: 'P2000',
  meta: { column_name: '<unknown>' }
```

**Comment by louisrli** [comment]
ah never mind -- i see a way to get the query params from the events in the documentation, let me try it

**Comment by louisrli** [comment]
```
prisma:query INSERT INTO "default$default"."RawContent" ("content","id") VALUES ($1,$2) RETURNING "default$default"."RawContent"."id"
["Adrenal cortex releases the corticoids: <c>aldosterone</c> (mineralocorticoids) and <c>cortisol</c> (glucocorticoids).
Also releases <c>androgens</c> which are converted to testosterone and estrogen in gonads.","ckc5xsk7k10002wsbhy0ip8t3s"]
```

```
>>> len("ckc5xsk7k10002wsbhy0ip8t3s")
26
```

:(

**Comment by janpio** [comment]
Thank you, now we know the exact problem here and can be sure that this is the right thing to understand and fix.

**Comment by louisrli** [comment]
Do you have a time estimate for a release on this? Debating whether to alter all my FK and PK columns to be 50 characters or something or whether to just wait for the release.

**Comment by janpio** [comment]
We are looking into this right now. 

There are basically two possible outcomes: 
We change our cuid implementation to be only 25 characters, then you can keep your columns unchanged. Or we figure out that >25 characters is actually the expected thing, then we would need to change the column to e.g. be `varchar(30)`. We would want to put that into the upgrade tool here - but we also need to figure out if changing the column type has negative aftermath for usage in Prisma 1, e.g. would `prisma1 deploy` undo it again. That would then define if this is a P1 compatible or a breaking change.

Will take a bit to figure this all out. (And if you decide to go at it yourself, be warned about the last bit with P1 compatibility if that is important for you.)

**Comment by matthewmueller** [comment]
Closing in favor of #43
